### ISSUE TITLE
OpenCL Kernel only takes up to 16 arguments

### ISSUE THREAD

**Issue by ViktorWalter** [opencv/opencv#8410]
##### System information (version)
- OpenCV => 3.2
- Operating System / Platform => Ubuntu 16.04
- Compiler => g++

##### Detailed description

cv::ocl::Kernel::args() only takes up to 16 arguments. This is hard-coded in the source (ocl.hpp). Presumably, the developer did not expect users to need more than that, but for my purposes, I need the option for more.
I would suggest either adding more overloads with more arguments (maybe a bit unflexible solution) or adding the option to add more arguments by calling the function again with some flag. The second option would require the index of the last added argument (i in the source) to be available from outside, so that the next call would be able to continue from that index, instead of overwriting previous arguments.

**Comment by csukuangfj** [comment]
You could use `Kernel::set` instead of `kernel::args` as a workaround.

**Comment by ViktorWalter** [comment]
Yeah, I can call Kernel::set for each individual argument, instead of using Kernel::args. This needs to be done for each argument starting with the first one (index = 0), and to keep the index the same way it is done in Kernel::args. Thanks for the idea.

**Comment by alalek** [comment]
> starting with the first on

Right, `index` here is OpenCL internal kernel index, not external argument from `Kernel::args()` (for example, UMat may push 2+ arguments (buffer handle, sizes, strides)).

**Comment by ViktorWalter** [comment]
I have just encountered a different, but related problem. The "step", that is normally inserted as an additional argument from UMat does not take into account the element size. The actual index of a pixel on coordinates X, Y in a matrix of type T then must be calculated as Y*(step/sizeof(T))+X+offset. In OpenCV 2.4 (I am migrating my code) this value was always inserted explicitly, usually as m.step/m.elemSize(). I am not sure if this is working as intended. Should I open a new issue?

**Comment by alalek** [comment]
step/offset is passed in bytes to OpenCL kernel. Using this avoids unnecessary multiplication instructions (you don't see them in OpenCL code, but they are here on low level)

Just pass data buffer as `uchar*` add offset/step calculations and cast to `__global T*`.
